src/kbucket.rs

```diff
@@ -115,34 +116,95 @@ impl KbucketTable {
                 }
             }
         }
-        SearchResult {
-            found: false,
-            bucket_index,
-            column_index: last_empty_index,
-        }
+        return Search::Failure(bucket_index, last_empty_index);
     }
 
-    fn find_bucket_index(&self, identifier: Identifier) -> usize {
+    fn xor_bucket_index(&self, identifier: Identifier) -> usize {
         let x = U256::from(self.local_node_id);
         let y = U256::from(identifier);
         let xor_distance = x ^ y;
 
-        let bucket_index = MAX_BUCKETS - (xor_distance.leading_zeros() as usize);
-        println!(
-            "Xor distance leading zeros, {}",
-            xor_distance.leading_zeros()
-        );
-        println!("Bucket index for given key: {}", bucket_index);
-        bucket_index
+        MAX_BUCKETS - ((xor_distance.leading_zeros() - 1) as usize)
     }
 }
 
-/// TODO:  Implement real deal tests!
-///
-/// Test find_node()      **Requires adding nodes to our table**
-/// Test search_table()
-/// Test add_node()
+// WIP:  Make better assertions
 #[cfg(test)]
 mod tests {
     use super::*;
+
+    fn mk_nodes(n: usize) -> (Node, Vec<Node>) {
+        let listen_addr = String::from("127.0.0.1").parse::<Ipv4Addr>().unwrap();
+        let port_start = 9000_u16;
+
+        let our_nodes: Vec<Node> = (0..n)
+            .into_iter()
+            .map(|i| mk_node(&listen_addr, port_start, i))
+            .collect();
+
+        if let Some((local_node, other_nodes)) = our_nodes.split_first() {
+            let other_nodes = other_nodes.to_vec();
+            (local_node.clone(), other_nodes.clone())
+        } else {
+            panic!("Not enough nodes were created");
+        }
+    }
+
+    fn mk_node(listen_addr: &Ipv4Addr, port_start: u16, index: usize) -> Node {
+        Node {
+            ip_address: listen_addr.clone(),
+            udp_port: port_start + index as u16,
+            node_id: [index as u8; 32],
+        }
+    }
+
+    #[test]
+    fn add_redundant_node() {
+        let (local_node, dummy_nodes) = mk_nodes(2);
+        let mut table = KbucketTable::new(local_node.node_id);
+
+        let result = table.add_node(dummy_nodes[0]);
+        assert!(result);
+        let result2 = table.add_node(dummy_nodes[0]);
+        assert!(!result2);
+    }
+
+    #[test]
+    fn find_node_present() {
+        let (local_node, dummy_nodes) = mk_nodes(5);
+        let mut table = KbucketTable::new(local_node.node_id);
+        for i in 1..dummy_nodes.len() {
+            table.add_node(dummy_nodes[i]);
+        }
```

Comment:
I would just tier the `nodes` themselves

and generally I'd change `add_node` to take a `&Node`, rather than a `Node` directly

```suggestion
        for node in dummy_nodes {
            table.add_node(node);
        }
```

Reply:
On it.  What's the reasoning behind passing a reference to .add_node() instead of passing a node directly?

Reply:
if the type is big enough it will be better to just pass a reference, rather than try to push through a bunch of data into the CPU caches

and as a rule of thumb if it is not clear from the type directly (e.g. u64 is clearly only 64 bits whereas I don't know about `Node` just from looking at it) then I just keep it simply and pass by reference where I can